### DISCUSSION TITLE
Add VisualInstance attribute to MeshLibrary items

### DISCUSSION THREAD

**Discussion by escapade-games** [godotengine/godot-proposals#5438] [3D]
This is a proposal created mainly to address issue [#40245](https://github.com/godotengine/godot/issues/40245), which if solved would help my game immensely.

What I would like to propose is that we add a VisualInstance attribute to MeshLibrary items.

Currently, each MeshLibrary item only has the following attributes:
<img width="133" alt="MeshLibrary item attributes" src="https://user-images.githubusercontent.com/91832396/190720659-8d31d96e-e3d4-4a62-bce2-9c75424bc803.png">

What I suggest is adding a VisualInstance attribute among them. Since we create MeshLibraries by converting a scene of MeshInstances, and MeshInstance already has the VisualInstance attribute, why not preserve the attribute when converting the scene to a MeshLibrary?

Not only would this solve the issue of all the GridMap tiles being visible only on cull_mask layer 1, but it also has the added flexibility of having different tiles in the same GridMap be visible on different VisualLayers.

I see that this would help not just me, but also many others facing this problem when using GridMaps, as evidenced from how this bug report was even made in the first place. Thus I hope you can consider this proposal.

**Comment by theraot** [comment]
Sounds like what you need is the `layers` property, not a `VisualInstance` per se. For reference `VisualInstance` is a base class for multiple nodes including `MeshInstance`. `GridMap` will be creating some `VisualInstance`s for you (`MultiMeshInstance`s actually). Thus, in theory it is a matter of copying the value along.

**Comment by jackdarker** [comment]
I tried to create a minimap by rendering just the wall-tiles from my levels gridmap to a subviewport. I thought, would just have to configure the viewport-camera culling layer and meshinstance3d visualinstance-layer correctly and be done with it. It took me some hours research why this is not working (due the lack of gridmap respecting setting of meshinstance3d visualinstance-layer configuration).
Yes, it is noted in the documentation. But it would be of much more help if this open issue gets solved after all that years.

**Reply by Calinou** [reply]
There's a PR aiming to address this:

- https://github.com/godotengine/godot/pull/94453
